### DISCUSSION TITLE
Vulnerability's in vaultwarden/server:latest but not in vaultwarden/server:alpine

### DISCUSSION THREAD

**Discussion by ndonkersloot** [dani-garcia/vaultwarden#2205] [Q&A] [answered]
Hi there!

First of all, I love vaultwarden. Am using it for quite some time now, actively following the github page and think you guys rock!

This year I want to focus on docker security myself and therefore started scanning images I use using tools like grype, trivy or clair.
When scanning the vaultwarden/server:latest image with trivy there are numerous vulnerability's found. When scanning the vaultwarden/server:alpine image with trivy there are no vulnerability's found. Whether these vulnerability's are exploitable or not is a discussion for another time.

My question is;
Alpine obviously has a lot less attack surface because Debian has much more packages installed.
Is there a downside of using Alpine, and why isn't Alpine the default?
Based on these results alpine seems much more preferable.

```
vaultwarden/server:latest (debian 10.11)
=========================================
Total: 319 (UNKNOWN: 2, LOW: 191, MEDIUM: 64, HIGH: 54, CRITICAL: 8)
```

```
vaultwarden/server:alpine (alpine 3.14.3)
=========================================
Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)
```

**Comment by BlackDex** [comment, marked as answer]
There was a downside of Alpine not supporting all Database. Which has been solved recently. Same goes for the supported platforms.

Also Alpine is using a different Memory Allocation because the usage of the Musl Libc instead of the GNU version which is slower and thus makes Vaultwarden on Alpine also a bit slower compared to the Debian variant.

Besides that, I don't know how those scanners work, and if they check patch levels instead of specific version numbers or whatever. So some vulns could be false positive. Also, the Debian image used is still buster based and not Bullseye because of some runtime issues, which will probably lower the amount of vulns found.

As you stated your self, not all vulns found or maybe even none could pose a security risk because it's never used or not able to be used as an attack surface via Vaultwarden as a running service.

Since people want to choose in any case, and some prefer Debian, some prefer Alpine, it will probably never going to be a good default for everybody.

Also, since we just started with the Alpine images being able to run all Databases and all Platforms we already were supporting via Debian it probably is currently still the better choice as a default. And, switching the current default could break current setups. So i do not think we will change it in the near future to be Alpine as a default.

I hope this answers your questions 😃.

**Reply by ndonkersloot** [reply]
Thank you very much for your detailed response! This answers all questions I had.
